### ISSUE TITLE
Feedback needed: ok to drop the tsc checker in favour of tsuquyomi?

### ISSUE THREAD

**Issue by lcd047** [scrooloose/syntastic#1811]
Checker `tsc` has received some attention lately, several people asking for it to support looking up for a `tsconfig.json` file, and generally expressing their discontent with the way the checker works in syntastic (cf. #1816, #1807, #1786, #1778, #1734, #1628, #899, and maybe others I can't find right now).  On the other hand, some of the same people have recommended [tsuquyomi][tsuquyomi] as a better, smoother alternative to the `tsc` checker.  Since checkers maintained by people who care about them are generally in a better position to adapt to people's needs, I propose dropping the internal `tsc` checker in favour of [tsuquyomi][tsuquyomi].

Unless you have major objections to this move, I'll go ahead with the change in about two weeks (counting from the date of this writing).

[tsuquyomi]: https://github.com/Quramy/tsuquyomi

**Comment by nomaed** [comment]
I just want to explain why `tsc` is problematic, if anyone doesn't know:

It is due to the fact that when `tsc` CLI receives a target file to transpile (the used command is `tsc --noEmit someScript.ts`), it completely ignores `tsconfig.json` by design. 

Even if using a lookup to find `tsconfig.json` in the directory tree and feeding the directory to `--project` argument, as long as a specific target file is given this will cause an error to be emitted: `error TS1148: Cannot compile modules unless the '--module' flag is provided.` (for example: https://github.com/Microsoft/TypeScript/issues/5497)

So basically, `tsc` can either work on a per-file basis (by specifying all options via CLI), or on a per-project basis, in which it is executed either with `--project <path-of-tsconfig-file>` or without any arguments (then `tsconfig.json` is automatically resolved from `pwd` upwards), in which case the build is performed according to the options from `tsconfig.json`.

The only way to make a plugin like syntastic work is locating and parsing `tsconfig.json` and building an equivalent CLI command for every `tsc` execution.

**Comment by lcd047** [comment]
@nomaed There has been ample discussion about the problems with `tsc` and the solutions suggested by syntastic for it in the threads I mentioned above.  There is even a [section in the FAQ][FAQ] dedicated to it.  If you have important additions to what has been said before please either add them to one of the existing threads, or post a new one.  I'd prefer this thread to focus on whether `tsc` should be dropped from syntastic or not.

[FAQ]: https://github.com/scrooloose/syntastic#faqconfig

**Comment by nomaed** [comment]
@lcd047 Alright, feel free to delete my comment.

**Comment by renstrom** [comment]
Just wanted to mention that Tsuquyomi also requires [Shougo/vimproc.vim](https://github.com/Shougo/vimproc.vim).

**Comment by lcd047** [comment]
Done in commit 6014bdc.  Thank you for the feedback.

**Comment by unclechu** [comment]
So, any alternatives? With which one I should replace `tsc` now?

**Comment by nomaed** [comment]
@unclechu I am using  [tsuquyomi](https://github.com/Quramy/tsuquyomi)

**Comment by lcd047** [comment]
Please see also the [FAQ](https://github.com/scrooloose/syntastic#faqtsc).

**Comment by unclechu** [comment]
@nomaed Any advise how I can configure auto build for 'Shougo/vimproc.vim'? Like this example for NeoBundle:
```
NeoBundle 'Shougo/vimproc.vim', {
\ 'build' : {
\     'windows' : 'tools\\update-dll-mingw',
\     'cygwin' : 'make -f make_cygwin.mak',
\     'mac' : 'make -f make_mac.mak',
\     'linux' : 'make',
\     'unix' : 'gmake',
\    },
\ }

NeoBundle 'Quramy/tsuquyomi'
```

**Comment by lcd047** [comment]
@unclechu Can you please take this to the apropriate forum?  Thank you!

**Comment by unclechu** [comment]
@lcd047 Okay, but I'll paste solution here for anyone who will be frustrated by seeing this issue.

**Comment by aminer-looker** [comment]
I'm new to TypeScript, and tried using tsuquyomi and have had nothing but problems (see Quramy/tsuquyomi#82).  So, I have, reluctantly, tried to revive support for `tsc` until the problems have been fixed.  You can find [my adapted copy](https://github.com/andrewminer/init-env/blob/master/home/.vim/syntax_checkers/typescript/tsc.vim) on GitHub as well.

The changes I made to work for my environment basically amount to always compiling the whole project instead of the individual file (i.e., using the `-p tsconfig.json` file), and always using `--noEmit` on the output (since that option has now been around for a while).

Anyone who, like me, lands here from Google is welcome to use my file to try reverting to `tsc`.
